File: typescript/graph.ts
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// Copyright (c) 2026, Victor Duarte

import { DuplicateLinkError, GraphDefinitionError, RunError } from './contracts.js'
import { startRuntime } from './runner.js'
import { captureInitialState, resolveState } from './state.js'

import type {
  Action,
  FlowCombineHandler,
  FlowRecoveryHandler,
  NodeHandler,
  NodeRecoveryHandler,
  NormalizedRetryPolicy,
  RetryPolicy,
  RunHandle,
} from './contracts.js'

const stateInvariant: unique symbol = Symbol('caskada.state')

export interface Link<State extends object = Record<string, unknown>> {
  readonly action: Action | null
  readonly target: GraphElement<State>
}

export abstract class GraphElement<State extends object = Record<string, unknown>> {
  declare private readonly [stateInvariant]: (state: State) => State
  readonly name: string
  readonly #links = new Map<Action | null, Link<State>>()

  protected constructor(name: string) {
    this.name = nonemptyString(name, 'element name')
  }

  link(target: GraphElement<State>): void
  link(target: GraphElement<State>, action: Action): void
  link(target: GraphElement<State>, action?: Action): void {
    if (!(target instanceof GraphElement)) throw new GraphDefinitionError('link target must be a Node or Flow')
    const key = arguments.length === 1 ? null : nonemptyString(action, 'link action')
    if (this.#links.has(key)) {
      throw new DuplicateLinkError(`duplicate link action: ${key === null ? 'unlabelled' : JSON.stringify(key)}`)
    }
    this.#links.set(key, Object.freeze({ action: key, target }))
  }

  links(): readonly Link<State>[] {
    return Object.freeze([...this.#links.values()])
  }
}

export interface NodeOptions<State extends object = Record<string, unknown>, Input = unknown> {
  readonly name?: string
  readonly retry?: RetryPolicy
  readonly recover?: NodeRecoveryHandler<State, Input>
}

export class Node<State extends object = Record<string, unknown>, Input = unknown> extends GraphElement<State> {
  constructor(handler: NodeHandler<State, Input>, options: NodeOptions<State, Input> = {}) {
    if (typeof handler !== 'function') throw new GraphDefinitionError('node handler must be a function')
    const captured = optionsRecord(options, ['name', 'retry', 'recover'], 'Node')
    const inferredName = typeof handler.name === 'string' && handler.name ? handler.name : 'anonymous'
    const name = captured.name === undefined ? inferredName : nonemptyString(captured.name, 'Node.name')
    const recover = optionalFunction<NodeRecoveryHandler<State, Input>>(captured.recover, 'Node.recover')
    super(name)
    nodeData.set(this, {
      handler,
      retry: normalizeRetry(captured.retry),
      recover,
    })
  }
}

export function node<State extends object = Record<string, unknown>, Input = unknown>(
  handler: NodeHandler<State, Input>,
  options: NodeOptions<State, Input> = {},
): Node<State, Input> {
  return new Node(handler, options)
}

export interface FlowOptions<State extends object = Record<string, unknown>> {
  readonly name?: string
  readonly exits?: readonly Action[]
  readonly concurrency?: number
  readonly maxActivations?: number
  readonly combine?: FlowCombineHandler<State>
  readonly recover?: FlowRecoveryHandler<State>
}

export class Flow<State extends object = Record<string, unknown>> extends GraphElement<State> {
  constructor(entry: GraphElement<State>, options: FlowOptions<State> = {}) {
    if (!(entry instanceof GraphElement)) throw new GraphDefinitionError('Flow.entry must be a Node or Flow')
    const captured = optionsRecord(options, ['name', 'exits', 'concurrency', 'maxActivations', 'combine', 'recover'], 'Flow')
    const name = captured.name === undefined ? 'Flow' : nonemptyString(captured.name, 'Flow.name')
    const exits = captureExits(captured.exits)
    const concurrency = captured.concurrency === undefined ? 1 : positiveInteger(captured.concurrency, 'Flow.concurrency')
    const maxActivations =
      captured.maxActivations === undefined ? undefined : positiveInteger(captured.maxActivations, 'Flow.maxActivations')
    const combine = optionalFunction<FlowCombineHandler<State>>(captured.combine, 'Flow.combine')
    const recover = optionalFunction<FlowRecoveryHandler<State>>(captured.recover, 'Flow.recover')
    super(name)
    flowData.set(this, { entry, exits, concurrency, maxActivations, combine, recover })
  }

  get entry(): GraphElement<State> {
    return requireData(flowData, this, 'Flow').entry
  }

  get exits(): readonly Action[] {
    return requireData(flowData, this, 'Flow').exits
  }

  get concurrency(): number {
    return requireData(flowData, this, 'Flow').concurrency
  }

  get maxActivations(): number | undefined {
    return requireData(flowData, this, 'Flow').maxActivations
  }

  compile(): CompiledFlow<State> {
    return new CompiledFlowImpl(compile(this))
  }

  start(initialState: Readonly<State>): RunHandle<State> {
    return this.compile().start(initialState)
  }

  run(initialState: Readonly<State>): Promise<State> {
    return this.compile().run(initialState)
  }
}

export interface CompiledFlow<State extends object = Record<string, unknown>> {
  start(initialState: Readonly<State>): RunHandle<State>
  run(initialState: Readonly<State>): Promise<State>
  describe(): CompiledDescription
}

class CompiledFlowImpl<State extends object> implements CompiledFlow<State> {
  readonly #snapshot: CompiledSnapshot

  constructor(snapshot: CompiledSnapshot) {
    this.#snapshot = snapshot
  }

  start(initialState: Readonly<State>): RunHandle<State> {
    return startRuntime(this.#snapshot, captureInitialState(initialState))
  }

  run(initialState: Readonly<State>): Promise<State> {
    const handle = this.start(initialState)
    return new Promise<State>((resolve, reject) => {
      handle.result().then((result) => {
        if (result.status === 'failed') {
          reject(new RunError(result))
        } else {
          resolveState(resolve, result.state)
        }
      }, reject)
    })
  }

  describe(): CompiledDescription {
    return describe(this.#snapshot)
  }
}

export interface CompiledLink {
  readonly action: Action | null
  readonly targetElementId: number
}

export interface CompiledPlacement {
  readonly elementId: number
  readonly kind: 'node' | 'flow'
  readonly name: string
  readonly links: readonly CompiledLink[]
  readonly handler?: NodeHandler<any, any>
  readonly retry?: NormalizedRetryPolicy
  readonly recover?: NodeRecoveryHandler<any, any>
  readonly ownedScopeId?: number
}

export interface CompiledScope {
  readonly scopeId: number
  readonly ownerElementId: number
  readonly parentScopeId: number | null
  readonly entryElementId: number
  readonly name: string
  readonly exits: readonly Action[]
  readonly concurrency: number
  readonly maxActivations: number | undefined
  readonly combine: FlowCombineHandler<any> | undefined
  readonly recover: FlowRecoveryHandler<any> | undefined
}

export interface CompiledSnapshot {
  readonly scopes: readonly CompiledScope[]
  readonly placements: readonly CompiledPlacement[]
}

export interface CompiledDescription {
  readonly schema_version: 1
  readonly root: { readonly element_id: 1; readonly scope_id: 1 }
  readonly scopes: readonly Record<string, unknown>[]
  readonly elements: readonly Record<string, unknown>[]
}

interface NodeData {
  readonly handler: NodeHandler<any, any>
  readonly retry: NormalizedRetryPolicy
  readonly recover: NodeRecoveryHandler<any, any> | undefined
}

interface FlowData {
  readonly entry: GraphElement<any>
  readonly exits: readonly Action[]
  readonly concurrency: number
  readonly maxActivations: number | undefined
  readonly combine: FlowCombineHandler<any> | undefined
  readonly recover: FlowRecoveryHandler<any> | undefined
}

const nodeData = new WeakMap<Node<any, any>, NodeData>()
const flowData = new WeakMap<Flow<any>, FlowData>()

interface ScopeWork {
  readonly scopeId: number
  readonly ownerElementId: number
  readonly parentScopeId: number | null
  readonly flow: Flow<any>
  readonly ancestors: readonly Flow<any>[]
}

class Compiler {
  readonly placements = new Map<number, CompiledPlacement>()
  readonly scopes = new Map<number, CompiledScope>()
  readonly work: ScopeWork[]
  readonly ownedScopes = new Map<number, number>()
  nextElementId = 2
  nextScopeId = 2

  constructor(root: Flow<any>) {
    this.placements.set(1, { elementId: 1, kind: 'flow', name: root.name, links: [], ownedScopeId: 1 })
    this.ownedScopes.set(1, 1)
    this.work = [{ scopeId: 1, ownerElementId: 1, parentScopeId: null, flow: root, ancestors: [root] }]
  }

  compile(): CompiledSnapshot {
    for (const scope of this.work) this.compileScope(scope)
    return Object.freeze({
      scopes: Object.freeze(Array.from({ length: this.nextScopeId - 1 }, (_, index) => required(this.scopes, index + 1))),
      placements: Object.freeze(Array.from({ length: this.nextElementId - 1 }, (_, index) => required(this.placements, index + 1))),
    })
  }

  private compileScope(scope: ScopeWork): void {
    const ids = new Map<GraphElement<any>, number>()
    const elements: GraphElement<any>[] = []

    const add = (element: GraphElement<any>): number => {
      const existing = ids.get(element)
      if (existing !== undefined) return existing
      if (!(element instanceof Node) && !(element instanceof Flow)) {
        throw new GraphDefinitionError('unsupported graph element')
      }
      const elementId = this.nextElementId++
      ids.set(element, elementId)
      elements.push(element)
      if (element instanceof Flow) {
        if (scope.ancestors.includes(element)) {
          throw new GraphDefinitionError('recursive Flow containment is not allowed')
        }
        const ownedScopeId = this.nextScopeId++
        this.ownedScopes.set(elementId, ownedScopeId)
        this.work.push({
          scopeId: ownedScopeId,
          ownerElementId: elementId,
          parentScopeId: scope.scopeId,
          flow: element,
          ancestors: [...scope.ancestors, element],
        })
      }
      return elementId
    }

    const entryElementId = add(requireData(flowData, scope.flow, 'Flow').entry)
    for (const element of elements) {
      const links = Object.freeze(element.links().map((link) => ({ action: link.action, targetElementId: add(link.target) })))
      const elementId = required(ids, element)
      if (element instanceof Node) {
        const data = requireData(nodeData, element, 'Node')
        this.placements.set(elementId, {
          elementId,
          kind: 'node',
          name: element.name,
          links,
          handler: data.handler,
          retry: data.retry,
          ...(data.recover === undefined ? {} : { recover: data.recover }),
        })
      } else {
        this.placements.set(elementId, {
          elementId,
          kind: 'flow',
          name: element.name,
          links,
          ownedScopeId: required(this.ownedScopes, elementId),
        })
      }
    }

    const data = requireData(flowData, scope.flow, 'Flow')
    this.scopes.set(scope.scopeId, {
      scopeId: scope.scopeId,
      ownerElementId: scope.ownerElementId,
      parentScopeId: scope.parentScopeId,
      entryElementId,
      name: scope.flow.name,
      exits: data.exits,
      concurrency: data.concurrency,
      maxActivations: data.maxActivations,
      combine: data.combine,
      recover: data.recover,
    })
  }
}

function compile(root: Flow<any>): CompiledSnapshot {
  return new Compiler(root).compile()
}

function describe(snapshot: CompiledSnapshot): CompiledDescription {
  return {
    schema_version: 1,
    root: { element_id: 1, scope_id: 1 },
    scopes: snapshot.scopes.map((scope) => ({
      scope_id: scope.scopeId,
      owner_element_id: scope.ownerElementId,
      parent_scope_id: scope.parentScopeId,
      entry_element_id: scope.entryElementId,
      name: scope.name,
      exits: [...scope.exits],
      concurrency: scope.concurrency,
      max_activations: scope.maxActivations ?? null,
    })),
    elements: snapshot.placements.map((element) => ({
      element_id: element.elementId,
      kind: element.kind,
      name: element.name,
      links: element.links.map((link) => ({ action: link.action, target_element_id: link.targetElementId })),
      ...(element.retry === undefined ? { owned_scope_id: element.ownedScopeId } : { max_attempts: element.retry.maxAttempts }),
    })),
  }
}

function normalizeRetry(value: unknown): NormalizedRetryPolicy {
  const retry = optionsRecord(value ?? {}, ['maxAttempts', 'shouldRetry', 'delayMs'], 'RetryPolicy')
  const maxAttempts = retry.maxAttempts === undefined ? 1 : positiveInteger(retry.maxAttempts, 'RetryPolicy.maxAttempts')
  const shouldRetry =
    retry.shouldRetry === undefined
      ? () => true
      : optionalFunction<(failure: any) => boolean>(retry.shouldRetry, 'RetryPolicy.shouldRetry')
  const delayMs =
    retry.delayMs === undefined
      ? 0
      : typeof retry.delayMs === 'function'
        ? retry.delayMs
        : nonnegativeInteger(retry.delayMs, 'RetryPolicy.delayMs')
  return Object.freeze({ maxAttempts, shouldRetry: shouldRetry!, delayMs })
}

function captureExits(value: unknown): readonly string[] {
  if (value === undefined) return Object.freeze([])
  if (!Array.isArray(value)) throw new GraphDefinitionError('Flow.exits must be an array of actions')
  const exits = value.map((action) => nonemptyString(action, 'Flow exit'))
  if (new Set(exits).size !== exits.length) throw new GraphDefinitionError('Flow.exits contains a duplicate action')
  return Object.freeze(exits)
}

function optionsRecord(value: unknown, allowed: readonly string[], name: string): Record<string, any> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new GraphDefinitionError(`${name} options must be an object`)
  }
  for (const key of Object.keys(value)) {
    if (!allowed.includes(key)) throw new GraphDefinitionError(`unknown ${name} option: ${key}`)
  }
  return value as Record<string, any>
}

function optionalFunction<T>(value: unknown, field: string): T | undefined {
  if (value === undefined) return undefined
  if (typeof value !== 'function') throw new GraphDefinitionError(`${field} must be a function`)
  return value as T
}

function nonemptyString(value: unknown, field: string): string {
  if (typeof value !== 'string' || !value) throw new GraphDefinitionError(`${field} must be a nonempty string`)
  return value
}

function positiveInteger(value: unknown, field: string): number {
  if (!Number.isSafeInteger(value) || (value as number) < 1) {
    throw new GraphDefinitionError(`${field} must be a positive safe integer`)
  }
  return value as number
}

function nonnegativeInteger(value: unknown, field: string): number {
  if (!Number.isSafeInteger(value) || (value as number) < 0) {
    throw new GraphDefinitionError(`${field} must be a nonnegative safe integer`)
  }
  return value as number
}

function requireData<K extends object, V>(map: WeakMap<K, V>, key: K, name: string): V {
  const value = map.get(key)
  if (value === undefined) throw new Error(`${name} data is missing`)
  return value
}

function required<K, V>(map: Map<K, V>, key: K): V {
  const value = map.get(key)
  if (value === undefined) throw new Error('compiled graph data is missing')
  return value
}

File: python/caskada/_graph.py
# This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
# Copyright (c) 2026, Victor Duarte
from __future__ import annotations

from collections.abc import Callable, Iterable
from dataclasses import dataclass
from typing import Any, Generic, Literal, TypeAlias, cast, overload

from ._contracts import (
    Action,
    DuplicateLinkError,
    FlowCombineHandler,
    FlowRecoveryHandler,
    GraphDefinitionError,
    InputT,
    NodeHandler,
    NodeRecoveryHandler,
    RetryPolicy,
    RunError,
    RunHandle,
    StateT,
    _nonempty_string,
    _positive_integer,
)
from ._state import _capture_initial_state

_UNLABELLED = object()
_MISSING = object()
CompiledDescription: TypeAlias = dict[str, object]


@dataclass(frozen=True, slots=True)
class Link(Generic[StateT]):
    action: Action | None
    target: GraphElement[StateT]


class GraphElement(Generic[StateT]):
    def __init__(self, name: str) -> None:
        self._name = _nonempty_string(name, "element name")
        self._links: dict[object, Link[StateT]] = {}

    @property
    def name(self) -> str:
        return self._name

    @overload
    def link(self, target: GraphElement[StateT], /) -> None: ...

    @overload
    def link(self, target: GraphElement[StateT], action: Action, /) -> None: ...

    def link(
        self,
        target: GraphElement[StateT],
        action: Action | object = _UNLABELLED,
        /,
    ) -> None:
        if not isinstance(target, GraphElement):
            raise GraphDefinitionError("link target must be a Node or Flow")
        key: object = _UNLABELLED
        public_action: str | None = None
        if action is not _UNLABELLED:
            public_action = _nonempty_string(action, "link action")
            key = public_action
        if key in self._links:
            label = "unlabelled" if key is _UNLABELLED else repr(public_action)
            raise DuplicateLinkError(f"duplicate link action: {label}")
        self._links[key] = Link(public_action, target)

    def links(self) -> tuple[Link[StateT], ...]:
        return tuple(self._links.values())


class Node(GraphElement[StateT], Generic[StateT, InputT]):
    def __init__(
        self,
        handler: NodeHandler[StateT, InputT],
        *,
        name: str,
        retry: RetryPolicy,
        recover: NodeRecoveryHandler[StateT, InputT] | None,
    ) -> None:
        super().__init__(name)
        self.handler = handler
        self.retry = retry
        self.recover = recover


@overload
def node(
    handler: NodeHandler[StateT, InputT],
    /,
    *,
    name: str | None = None,
    retry: RetryPolicy | None = None,
    recover: NodeRecoveryHandler[StateT, InputT] | None = None,
) -> Node[StateT, InputT]: ...


@overload
def node(
    *,
    name: str | None = None,
    retry: RetryPolicy | None = None,
    recover: NodeRecoveryHandler[StateT, InputT] | None = None,
) -> Callable[[NodeHandler[StateT, InputT]], Node[StateT, InputT]]: ...


def node(
    handler: NodeHandler[StateT, InputT] | object = _MISSING,
    /,
    *,
    name: str | None = None,
    retry: RetryPolicy | None = None,
    recover: NodeRecoveryHandler[StateT, InputT] | None = None,
) -> (
    Node[StateT, InputT] | Callable[[NodeHandler[StateT, InputT]], Node[StateT, InputT]]
):
    policy = RetryPolicy() if retry is None else retry
    if not isinstance(policy, RetryPolicy):
        raise GraphDefinitionError("node retry must be a RetryPolicy")
    if recover is not None and not callable(recover):
        raise GraphDefinitionError("Node.recover must be callable")
    if name is not None:
        _nonempty_string(name, "Node.name")

    def create(callback: NodeHandler[StateT, InputT]) -> Node[StateT, InputT]:
        if not callable(callback):
            raise GraphDefinitionError("node handler must be callable")
        inferred = getattr(callback, "__name__", "anonymous")
        resolved_name = name or (inferred if isinstance(inferred, str) else "anonymous")
        return Node(callback, name=resolved_name, retry=policy, recover=recover)

    if handler is _MISSING:
        return create
    return create(cast(NodeHandler[StateT, InputT], handler))


class Flow(GraphElement[StateT], Generic[StateT]):
    def __init__(
        self,
        entry: GraphElement[StateT],
        *,
        name: str = "Flow",
        exits: Iterable[Action] = (),
        concurrency: int = 1,
        max_activations: int | None = None,
        combine: FlowCombineHandler[StateT] | None = None,
        recover: FlowRecoveryHandler[StateT] | None = None,
    ) -> None:
        if not isinstance(entry, GraphElement):
            raise GraphDefinitionError("Flow.entry must be a Node or Flow")
        if isinstance(exits, (str, bytes)):
            raise GraphDefinitionError("Flow.exits must be an iterable of actions")
        captured_exits = tuple(
            _nonempty_string(action, "Flow exit") for action in exits
        )
        if len(set(captured_exits)) != len(captured_exits):
            raise GraphDefinitionError("Flow.exits contains a duplicate action")
        _positive_integer(concurrency, "Flow.concurrency")
        if max_activations is not None:
            _positive_integer(max_activations, "Flow.max_activations")
        if combine is not None and not callable(combine):
            raise GraphDefinitionError("Flow.combine must be callable")
        if recover is not None and not callable(recover):
            raise GraphDefinitionError("Flow.recover must be callable")

        super().__init__(name)
        self.entry = entry
        self.exits = captured_exits
        self.concurrency = concurrency
        self.max_activations = max_activations
        self.combine = combine
        self.recover = recover

    def compile(self) -> CompiledFlow[StateT]:
        return CompiledFlow(_compile(self))

    def start(self, initial_state: StateT) -> RunHandle[StateT]:
        return self.compile().start(initial_state)

    async def run(self, initial_state: StateT) -> StateT:
        return await self.compile().run(initial_state)


class CompiledFlow(Generic[StateT]):
    def __init__(self, snapshot: _CompiledSnapshot) -> None:
        self._snapshot = snapshot

    def start(self, initial_state: StateT) -> RunHandle[StateT]:
        from ._runner import _start

        return _start(self._snapshot, _capture_initial_state(initial_state))

    async def run(self, initial_state: StateT) -> StateT:
        result = await self.start(initial_state).result()
        if result.status == "failed":
            raise RunError(result)
        return result.state

    def describe(self) -> CompiledDescription:
        return _describe(self._snapshot)


@dataclass(frozen=True, slots=True)
class _CompiledLink:
    action: Action | None
    target_element_id: int


@dataclass(frozen=True, slots=True)
class _CompiledPlacement:
    element_id: int
    kind: Literal["node", "flow"]
    name: str
    links: tuple[_CompiledLink, ...]
    handler: Callable[..., Any] | None = None
    retry: RetryPolicy | None = None
    recover: Callable[..., Any] | None = None
    owned_scope_id: int | None = None


@dataclass(frozen=True, slots=True)
class _CompiledScope:
    scope_id: int
    owner_element_id: int
    parent_scope_id: int | None
    entry_element_id: int
    name: str
    exits: tuple[Action, ...]
    concurrency: int
    max_activations: int | None
    combine: Callable[..., Any] | None
    recover: Callable[..., Any] | None


@dataclass(frozen=True, slots=True)
class _CompiledSnapshot:
    scopes: tuple[_CompiledScope, ...]
    placements: tuple[_CompiledPlacement, ...]

    def scope(self, scope_id: int) -> _CompiledScope:
        return self.scopes[scope_id - 1]

    def placement(self, element_id: int) -> _CompiledPlacement:
        return self.placements[element_id - 1]


@dataclass(slots=True)
class _ScopeWork:
    scope_id: int
    owner_element_id: int
    parent_scope_id: int | None
    flow: Flow[Any]
    ancestors: tuple[Flow[Any], ...]


class _Compiler:
    def __init__(self, root: Flow[Any]) -> None:
        self.next_element_id = 2
        self.next_scope_id = 2
        self.placements: dict[int, _CompiledPlacement] = {
            1: _CompiledPlacement(1, "flow", root.name, (), owned_scope_id=1)
        }
        self.scopes: dict[int, _CompiledScope] = {}
        self.work = [_ScopeWork(1, 1, None, root, (root,))]

    def compile(self) -> _CompiledSnapshot:
        for scope in self.work:
            self._compile_scope(scope)
        return _CompiledSnapshot(
            tuple(self.scopes[index] for index in range(1, self.next_scope_id)),
            tuple(self.placements[index] for index in range(1, self.next_element_id)),
        )

    def _compile_scope(self, scope: _ScopeWork) -> None:
        element_ids: dict[GraphElement[Any], int] = {}
        elements: list[GraphElement[Any]] = []

        def add(element: GraphElement[Any]) -> int:
            existing = element_ids.get(element)
            if existing is not None:
                return existing
            if not isinstance(element, (Node, Flow)):
                raise GraphDefinitionError("unsupported graph element")
            element_id = self.next_element_id
            self.next_element_id += 1
            element_ids[element] = element_id
            elements.append(element)
            if isinstance(element, Flow):
                if element in scope.ancestors:
                    raise GraphDefinitionError(
                        "recursive Flow containment is not allowed"
                    )
                owned_scope_id = self.next_scope_id
                self.next_scope_id += 1
                self.work.append(
                    _ScopeWork(
                        owned_scope_id,
                        element_id,
                        scope.scope_id,
                        element,
                        (*scope.ancestors, element),
                    )
                )
            return element_id

        entry_id = add(scope.flow.entry)
        for element in elements:
            links = tuple(
                _CompiledLink(link.action, add(link.target)) for link in element.links()
            )
            element_id = element_ids[element]
            if isinstance(element, Node):
                self.placements[element_id] = _CompiledPlacement(
                    element_id,
                    "node",
                    element.name,
                    links,
                    element.handler,
                    element.retry,
                    element.recover,
                )
            else:
                owned_scope_id = next(
                    item.scope_id
                    for item in self.work
                    if item.owner_element_id == element_id
                )
                self.placements[element_id] = _CompiledPlacement(
                    element_id,
                    "flow",
                    element.name,
                    links,
                    owned_scope_id=owned_scope_id,
                )

        self.scopes[scope.scope_id] = _CompiledScope(
            scope.scope_id,
            scope.owner_element_id,
            scope.parent_scope_id,
            entry_id,
            scope.flow.name,
            scope.flow.exits,
            scope.flow.concurrency,
            scope.flow.max_activations,
            scope.flow.combine,
            scope.flow.recover,
        )


def _compile(root: Flow[StateT]) -> _CompiledSnapshot:
    if not isinstance(root, Flow):
        raise GraphDefinitionError("compile requires a Flow")
    return _Compiler(root).compile()


def _describe(snapshot: _CompiledSnapshot) -> CompiledDescription:
    return {
        "schema_version": 1,
        "root": {"element_id": 1, "scope_id": 1},
        "scopes": [
            {
                "scope_id": scope.scope_id,
                "owner_element_id": scope.owner_element_id,
                "parent_scope_id": scope.parent_scope_id,
                "entry_element_id": scope.entry_element_id,
                "name": scope.name,
                "exits": list(scope.exits),
                "concurrency": scope.concurrency,
                "max_activations": scope.max_activations,
            }
            for scope in snapshot.scopes
        ],
        "elements": [
            {
                "element_id": element.element_id,
                "kind": element.kind,
                "name": element.name,
                "links": [
                    {
                        "action": link.action,
                        "target_element_id": link.target_element_id,
                    }
                    for link in element.links
                ],
                **(
                    {"max_attempts": element.retry.max_attempts}
                    if element.retry is not None
                    else {"owned_scope_id": element.owned_scope_id}
                ),
            }
            for element in snapshot.placements
        ],
    }
